### DISCUSSION TITLE
Document upgrade and downgrade processes

### DISCUSSION THREAD

**Discussion by zamazan4ik** [dexidp/dex#2621] [General]
Hi.

After reading the documentation it is not clear for me, how upgrade and downgrade processes should be performed over the solution. Could you please describe somewhere in the documentation:

- How to upgrade Dex to the newer version (hopefully with no downtime)
- How to downgrade Dex to the older version (hopefully with no downtime)
- Maybe some compatibility policies and notes between versions

Thanks in advance!

**Comment by thiDucTran** [comment]
i feel like we just change the `image:` but yeah..would be great to have official documentation

**Reply by nabokihms** [reply]
Dex is a single binary that works with external storage and providers running with a single config.

Usually, we do not introduce breaking changes, so:
1. Old config files work with new versions of Dex. If you start using a feature from a more recent version of Dex and want to downgrade, you must migrate the config back first.
2. Storage only has UP migrations, old Dex versions work with new storage. We have not introduced complicated storage migrations for years.

Regarding downtime-safe upgrades, we can recommend running an instance of a new Dex version, switching the traffic from the old instance to the new one, and terminating the old one. For example, Kubernetes and its Deployment controller do this out of the box with the `RollingUpdate` mode.

**Comment by nabokihms** [comment]
https://github.com/dexidp/website/issues/139
